Isabelle 2020 update

lsf37 opened the pull request:
This PR updates the `l4v` repo to Isabelle2020 (to be used with `ts-2020` branch of the isabelle repo).

The main notable points are:
  - Isabelle has become a bit slower (records more informations for some thms)
  - every session now has to has its own directory. This means there is some file moving going on in the update. The requirement flushed out a few last pieces where we relied on implicit paths, which is good. It also made it necessary to introduce a few more directories like `refine/base`, because `BaseRefine` needs its own directory now. After grumbling a while, I actually kinda like it now.
  - session `ASpecDoc` is merged back into `ASpec` (because of the session directories). I've come up with a mechanism that doesn't rebuild every time we have a new git revision. It still means you have to run `make ASpec` to get all files the session depends on (i.e. just `isabelle build ASpec` without `make` first will complain).
  - theory merge order seems to have changed slightly, mostly affecting global declaration status of `if_cong` and `option.case_cong`. Instead of trying to replicate the previous global declarations, I've made local (per thm) fixes for all breakages, which hopefully will provide stability against the next such change and also records where we actually depended on these. The downside is that there are many `supply if_cong[cong]` sprinkled over the proofs now.
  - `AutoLevityBase` depended on an internal simplifier API that is no longer exported. AFAICT it was used to record/trace `cong` rules. Instead of patching Isabelle, I've commented out the `cong` rule tracing in `AutoLevityBase` 
  - `SimplExportAndRefine` takes even longer than before. Somebody should look at where the time is going at some point.
  - `Word_Lib` is in sync with `AFP-2020`
  - the Isabelle monad syntax changed slightly (flipped associativity to what we're using, IIRC) -- very few changes, because we only use that syntax occasionally.  
  - apart from a few word tweaks and the `if_cong` sprinkle, very few actual proof updates

lsf37 commented:
Looks like I still have a few small gitlint/link things to fix up, but the theory files are ready for review. For gitlint I'll have to rebase/force push, so maybe give me another hour before looking at it.

lsf37 commented:
Testboard running [here](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-631)

lsf37 commented:
Oh, the github CI checks will all fail, of course, because the manifest is still on Isabelle2019/`ts-2019`. We'll have to trust the testboard for this one.

mbrcknl commented on proof/refine/ARM/Detype_R.thy:
```diff
@@ -3348,27 +3348,20 @@ lemma curDomain_commute:
-  done
+  by (auto simp: obj_at'_def is_aligned_mask lookupAround2_None1
+                    lookupAround2_char1 field_simps objBits_simps
+                    projectKO_opt_tcb projectKO_def return_def ps_clear_def
+           split: if_splits
+           dest!: pspace_no_overlap_disjoint')
```
💯

mbrcknl commented on proof/refine/ARM_HYP/KHeap_R.thy:
```diff
@@ -713,9 +702,9 @@ lemma cte_wp_at_ctes_of:
   apply (clarsimp split: if_split_asm del: disjCI)
    apply (simp add: ps_clear_def3 field_simps)
-  apply (rule disjI2, rule exI[where x="(p - (p && ~~ mask 9))"])
+  apply (rule disjI2, rule exI[where x="p - (p && ~~ mask 9)"])
```
I'm curious why this didn't need parens for `~~ (mask 9)`, when that seems to have been necessary in lots of other places?

lsf37 replied:
The other places with the many changes were synced from the AFP where that change was applied globally once, but I didn't undo it globally when I reverted the syntax change (the real change Florian wanted was for `NOT`, but doing the same for `~~` was a bit too much, the syntax precedence we have is analogous to other symbolic not-operators).

mbrcknl replied:
See @lsf37's reply to my comment on `lib/Word_Lib/Aligned.thy`, which answers this.

mbrcknl commented on spec/abstract/document/root.tex:
```diff
@@ -89,6 +88,36 @@
 
 \begin{document}
 
+% tex hacking to get git revision from repo document
+
+% we can't include gitrev.tex in document_files, because this will rebuild
+% ASpec (and therefore pretty much everything) every time the git revision
+% changes (for any change anywhere). This means gitrev.tex will not be
+% copied into the document build directory, and we have to find the repo
+% checkout. We give the root of the repo checkout as document file, because
+% if that changes, everything changes anyway. From there we can find the rest,
+% which we do below:
+
+\newcommand\gitroot{}
+\newcommand\gitrev{}
+
+\makeatletter\let\myfilehandle\@inputcheck\makeatother
+\openin\myfilehandle=git-root.tex\relax
+\begingroup\endlinechar-1
+  \global\read\myfilehandle to \gitroot
+\endgroup
+\closein\myfilehandle
+
+\makeatletter\let\myfilehandle\@inputcheck\makeatother
+\openin\myfilehandle=\gitroot/spec/abstract/document/gitrev.tex\relax
+\begingroup\endlinechar-1
+  \global\read\myfilehandle to \gitrev
+\endgroup
+\closein\myfilehandle
+
+% end tex hacking
+
+
```
Clever!

mbrcknl commented on spec/Makefile:
```diff
@@ -23,26 +23,31 @@ report-regression:
+ASPEC_GITREV_FILE=abstract/document/gitrev.tex
+GIT_ROOT_FILE=abstract/document/git-root.tex
 
-ASpecDoc: design-spec $(ASPEC_DOC_VERSION_FILE) $(ASPEC_DOC_GITREV_FILE)
+# NOTE: The abstract spec imports Events from Haskell hence the dependency
+ASpec: $(ASPEC_VERSION_FILE) $(GIT_ROOT_FILE) $(ASPEC_GITREV_FILE) design-spec
 
-$(ASPEC_DOC_VERSION_FILE): $(SEL4_VERSION)
+$(ASPEC_VERSION_FILE): $(SEL4_VERSION)
 	cp $< $@
 
-$(ASPEC_DOC_GITREV_FILE): .FORCE
+$(GIT_ROOT_FILE):
+	git rev-parse --show-toplevel > $@
+	perl -p -i -e "s/\n//" $@
+
+$(ASPEC_GITREV_FILE): .FORCE
 	git rev-parse --short=15 HEAD > $@ || echo unknown > $@
 	git diff --no-ext-diff --quiet || echo "*" >> $@
 	git diff --no-ext-diff --quiet --cached || echo "+" >> $@
```
For anyone curious, see the comments @lsf37 added in `spec/abstract/document/root.tex`.

mbrcknl commented on tools/asmrefine/testfiles/global_array_swap_gref.thy:
```diff
@@ -186,7 +186,14 @@ val debug_step_tac = ProveSimplToGraphGoals.debug_step_tac
 
 schematic_goal "PROP ?P"
   apply (tactic \<open>resolve_tac @{context} [init_thm] 1\<close>)
+  (* FIXME: this is debbuging code only, but the following execption still points at a problem
+            in the tactic code, not just to a broken proof.
+
+  exception THEORY raised (line 971 of "thm.ML"):
+  solve_constraints: bad theories for theorem
+
   apply (all \<open>(solves \<open>tactic \<open>HEADGOAL (debug_tac @{context})\<close>\<close>)?\<close>)
+  *)
```
- [ ] @mbrcknl make a Jira issue.

mbrcknl replied:
Issue noted at [VER-1369](https://sel4.atlassian.net/browse/VER-1369)

lsf37 replied:
I should have included the issue in the comment: there was already VER-1295, but that one doesn't link to the discussion here, so I've linked the two issues in Jira so we have both.

mbrcknl reviewed (approved):
Looks really good! It also looks like it was quite a lot of work, so many thanks for taking that on. Nevertheless, it was quite easy to review in full.

lsf37 commented on lib/Word_Lib/Aligned.thy:
```diff
@@ -786,7 +786,7 @@ proof -
 
 lemma is_aligned_neg_mask:
-  "m \<le> n \<Longrightarrow> is_aligned (x && ~~ mask n) m"
+  "m \<le> n \<Longrightarrow> is_aligned (x && ~~ (mask n)) m"
```
I reverted the syntax change for the `~~` operator in the AFP (but not for `NOT` where the motivation for that change came from), so it's Ok again to write `~~mask n`, but at that stage Florian had already update all occurrences in the AFP, and I didn't touch those again.

mbrcknl replied:
Right, that makes sense, and answers a question I had elsewhere where I saw `~~mask 9`.

lsf37 commented on lib/Word_Lib/Word_Lemmas.thy:
```diff
@@ -14,54 +14,6 @@ theory Word_Lemmas
-instantiation Enum.finite_1 :: len
-begin
-  definition "len_of_finite_1 (x :: Enum.finite_1 itself) \<equiv> (1 :: nat)"
-  instance
-    by (standard, auto simp: len_of_finite_1_def)
-end
-
-instantiation Enum.finite_2 :: len
-begin
-  definition "len_of_finite_2 (x :: Enum.finite_2 itself) \<equiv> (2 :: nat)"
-  instance
-    by (standard, auto simp: len_of_finite_2_def)
-end
-
-instantiation Enum.finite_3 :: len
-begin
-  definition "len_of_finite_3 (x :: Enum.finite_3 itself) \<equiv> (4 :: nat)"
-  instance
-    by (standard, auto simp: len_of_finite_3_def)
-end
-
-(* Provide wf and less_induct for word.
-   wf may be more useful in loop proofs, less_induct in recursion proofs. *)
-lemma word_less_wf: "wf {(a, b). a < (b :: ('a::len) word)}"
-  apply (rule wf_subset, rule wf_measure)
-  apply clarify
-  apply (subst in_measure)
-  apply (erule unat_mono)
-  done
-
-lemma word_less_induct:
-  "\<lbrakk> \<And>x::('a::len) word. (\<And>y. y < x \<Longrightarrow> P y) \<Longrightarrow> P x \<rbrakk> \<Longrightarrow> P a"
-  using word_less_wf  by induct blast
-
-instantiation word :: (len) wellorder
-begin
-instance by (intro_classes) (metis word_less_induct)
-end
-
-
```
Yes, that's right.

lsf37 commented on proof/crefine/ARM_HYP/AutoCorresTest.thy:
```diff
@@ -1,636 +0,0 @@
-(*
- * Copyright 2014, General Dynamics C4 Systems
- *
- * SPDX-License-Identifier: GPL-2.0-only
- *)
-
-(* Experimental AutoCorres proofs over CRefine: work in progress *)
-
-theory AutoCorresTest
-imports Refine_C
-begin
-
```
I should have a closer look at that.

lsf37 replied:
Ok, this was just overlooked. I think this happened because RISC-V got merged into the update somewhere in the middle after that theory had already been moved. Will be removed after the rebase.

mktnk3 commented on proof/tests.xml:
```diff
@@ -19,7 +19,7 @@
         <test name="AInvs" cpu-timeout="14400">make AInvs</test>
         <test name="BaseRefine">make BaseRefine</test>
-        <test name="Refine" cpu-timeout="14400">make Refine</test>
+        <test name="Refine" cpu-timeout="28800">make Refine</test>
```
great 👍

mktnk3 commented on proof/refine/ARM_HYP/KHeap_R.thy:
```diff
@@ -258,13 +256,6 @@ lemma updateObject_default_result:
   "(x, s'') \<in> fst (updateObject_default e ko p q n s) \<Longrightarrow> x = injectKO e"
   by (clarsimp simp add: updateObject_default_def in_monad)
 
-lemma ps_clear_upd':
-  "ksPSpace s y = Some v \<Longrightarrow>
-    ps_clear x n (s' \<lparr> ksPSpace := ksPSpace s(y \<mapsto> v')\<rparr>) = ps_clear x n s"
-  by (rule iffI | clarsimp elim!: ps_clear_domE | fastforce)+
-
-lemmas ps_clear_updE'[elim] = iffD2[OF ps_clear_upd', rotated]
-
```
This is interesting. So, in ARM_HYP we somehow have exactly the same thing defined as `ps_clear_upd`, and this is a duplicate, which you removed. I’d actually prefer not having ‘ at the end in other arches, but I guess this is not a high priority thing. Maybe one day :)

lsf37 replied:
Yes, that should probably be a separate thing. I'm not sure any more why I noticed the duplication here, it's possible that Isabelle complained about something or gave a warning.

mktnk3 reviewed (approved):
This is a huge PR. Thanks for going through this, Gerwin. I looked through mostly to understand what changes there are. I don’t have any important comments, just a few minor things. I wonder how we will find the change in cong rule declaration while working on our proofs.

lsf37 commented on lib/test/WPTutorial.thy:
```diff
@@ -207,11 +207,23 @@ throwing the extra assumption away - supplying the wp rule hoare_drop_imps
 helpful to crunch, crunch_simps and crunch_wps, which contain the rules we
 added here.
+
+Finally, this proof needs to simply the predicates underneath an
```
Well spotted! [](url)

lsf37 commented:
(DCO and gitlint will be failing now until I have fixed up all review comments and rebased + squashed)

lsf37 commented:
@corlewis @Xaphiosis Since we already have two reviews, it'd be Ok for you to skip. Let me know if you still want to take a closer look or if you're happy for me to go ahead and rebase+merge (after another testboard).

lsf37 commented:
Alright, fixed up and rebased. DCO etc should now come back green, as should this [testboard](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-638).
